rfc/rfc-76/rfc-76.md

```diff
@@ -0,0 +1,143 @@
+
+## Implementation
+
+With the ability of new file slice algorithm and compaction based on completion time, we can choose to
+
+### Schedule Compaction Plan
+The `EventTimeBasedCompactionStrategy` strategy is responsible to calculate the event time threshold for compaction plan.
+Implement a method to calculate the event time threshold:
+```java
+    // return the threshold as instant time format
+    public String calcEventTimeThreshold(String compactionInstant, HoodieWriteConfig config)
+```
+
+When generating compaction plan, we take the event time threshold as the next base instant time and save it to `CompactionOpertion::nextBaseInstantTime`.
+
+
+### Execute Compaction Plan
+Get `nextBaseInstantTime` from `CompactionOpertion`, write base file to the `nextBaseInstantTime`.
+
+Sample code in HoodieCompactor:
+```java
+    String nextBaseInstantTime = compactionOpertion.getNextBaseInstantTime();
+    result = executionHelper.writeFileAndGetWriteStats(compactionHandler, operation, nextBaseInstantTime, scanner, oldDataFileOpt);
+```
+
+Follow up:
+
+With the `EventTimeBasedCompactionStrategy` above, there can be data with event time after the threshold being compacted.
+For cases users want exact event time split (users don't want any data after threshold exist in RO table),
+we need to perform the log record filter in compaction execution based on the time threshold.
+ 
+To implement exact event time split, we should modify `HoodieMergedLogRecordScanner` adding new methods to get the remained log records, 
+likely naming `getRemainLogsIterator` and `getRemainRecords` corresponding to `HoodieMergedLogRecordScanner::iterator` and `HoodieMergedLogRecordScanner::getRecords`.
+And we have to rewrite those remained log records back to new log file.
+
+
+### Compute Data Freshness / Completion
+Given the definition below:
+
+| Query based table               | Data Completion Time                                   | Data Freshness Time            |
+|---------------------------------|--------------------------------------------------------|--------------------------------|
+| Snapshot table (RT table)       | all data before the time been ingested into hudi       | the latest data been ingested  |
+| Read Optimized table (RO table) | all data before the time been compacted into base file | the latest data been compacted |
+
+For RT table, the data completion and freshness can be simply extracted from `HoodieWriteStat::minEventTime` and `HoodieWriteStat::maxEventTime`. Now they are used to report metrics `commitLatencyInMs` and `commitFreshnessInMs`.
+
+For RO table, the data freshness time is the maxEventTime of last compaction, can be extracted from last commit metadata simply.
+But for the data completion time, we can not simply take the `minEventTime` of last compaction since it may be data after `minEventTime` uncompacted. So it depends on the earliest event time of all un-compacted log files. 
```

Comment:
> un-compacted log files. 

The compaction is not visible if it is not complete, so how the un-compacted log files play a role here.

Reply:
The Data Completion Time of RO table only update when the compaction completed, and saved to commit metadata. 
As Completion Time of RO table means all data before the time been compacted into base file, and the `DataCompletionTimeOfReadOptimizedTable = min { event time of un-compact log files } - 1`, so we need the un-compacted log files to determine the `DataCompletionTimeOfReadOptimizedTable`.

To get those un-compacted log files, 
1. load the latest fileSystemView, extract all log files from fileSlices naming to `allLogFiles`
2. get all compacted log files by last compaction execution, naming to `logFilesCompacted`
3. the un-compacted log files equal to `Set(allLogFiles) - Set(logFilesCompacted)`, naming to `logFilesUnCompacted`

Got the un-compacted log files, we can get the minimum instantTime from those log files, naming to `minInstantFromUnCompactedLogFiles`.
Then load timeline from `minInstantFromUnCompactedLogFiles`, deserialize each .deltacommit metadata, extract the `minEventTime` for those `logFilesUnCompacted`, the final `minEventTime - 1` is `DataCompletionTimeOfReadOptimizedTable`

Reply:
> As Completion Time of RO table means all data before the time been compacted into base file, and the DataCompletionTimeOfReadOptimizedTable = min { event time of un-compact log files } - 1, so we need the un-compacted log files to determine the DataCompletionTimeOfReadOptimizedTable.

Sorry I can not get this part.

Reply:
| Query based table               | Data Completion Time                                   | Data Freshness Time            |
|---------------------------------|--------------------------------------------------------|--------------------------------|
| Snapshot table (RT table)       | all data before the time been ingested into hudi       | the latest data been ingested  |
| Read Optimized table (RO table) | all data before the time been compacted into base file | the latest data been compacted |

The data completion time of RO table is defined as all data before the time been compacted into base file.  We need to infer this value from the minEventTime of the un-compacted log files. 

Is this confuse you?

Reply:
> value from the minEventTime of the un-compacted log files.

Are you talking about the source files of the compaction? Actually the write handle will collect the event time automically.

Reply:
No, I am not talking about the source files of the compaction. I am talking about the the log files out of the compaction.

> Actually the write handle will collect the event time automically.
Yeah, we can extract the max event time from the commit metadata. But we can not take the min event time from the metadata, as there may by log files out of the compaction with lower event time.

Reply:
The log files beyong the compaction instant time got greater completion time, but there is possibiliy they got smaller event time than the compaction instant time, for data compleness, you can not include any uncompacted logs in the ro table read view.